### ISSUE TITLE
Failing tests on 2017 Mac running Mojave

### ISSUE THREAD

**Issue by claui** [Homebrew/brew#9170]
# Bug report

**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem? **Non-issues: brew diy; openzfs libs in /usr/lib**
- [x] ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

I ran `brew tests` because the PR template says I have to.

## What happened (include command output)

Lots of tests fail reproducibly.

See command output: ~https://gist.github.com/claui/4265e4adb4847a2f0dcc49fa53665161~ Update 2021-01-07: https://gist.github.com/claui/86fc80e9e451a46c44f4fa244edd9ee8

## What you expected to happen

All the tests pass.

[January 2019 was the last known time](https://github.com/Homebrew/brew/pull/5636) when all `brew tests` passed for me. Not aware of any relevant changes to my system since then. Been running Mojave since late 2018.

## Step-by-step reproduction instructions (by running `brew` commands)

On my 2017 MacBook Pro running Mojave, with Homebrew installed to `/usr/local`:

```
$ brew tests
```

Note:

- Tests started getting flaky around April 2019, and test failures then [kept piling up over 18 months](https://github.com/Homebrew/brew/pulls?page=1&q=author%3Aclaui).

- There are even more failures when I run `brew tests` on another prefix. In order to keep this issue report focused, I ran the tests on the default prefix so these extra failures are not included.

- To capture both stdout and stderr in a single gist for this issue report, I actually ran `brew tests >${filename} 2>&1`.

## Output of `brew config` and `brew doctor` commands

brew config:

<pre>
HOMEBREW_VERSION: 2.5.11-20-gaf41431
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: af4143150da32097f751b1538a5b6f6958bb2175
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: fe767b0ad799e1b7d8acdf769c1c0c5c6c534342
Core tap last commit: 56 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BAT: set
HOMEBREW_CASK_OPTS: []
HOMEBREW_DEVELOPER: set
HOMEBREW_EDITOR: '/usr/local/bin/code' -w
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_NO_AUTO_UPDATE: set
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 11.0 build 1100
Git: 2.29.2 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 14.0.2, 12.0.2, 11.0.9, 10.0.2, 9, 1.8.0_272, 1.8.0_202
macOS: 10.14.6-x86_64
CLT: 11.3.1.0.1.1576735732
Xcode: 11.3.1
</pre>

brew doctor:

<pre>
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed with `brew diy`.
You should find replacements for the following formulae:
  kdumpd
  meld-executables
  arm-elf-binutils
  openzfs-executables
  arm-elf-musl
  arm-elf-gcc
  virtualbox-executables
  diy-dedrm

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libdiskmgt.1.dylib
  /usr/local/lib/libnvpair.1.dylib
  /usr/local/lib/libuutil.1.dylib
  /usr/local/lib/libzfs.2.dylib
  /usr/local/lib/libzfs_core.1.dylib
  /usr/local/lib/libzpool.1.dylib

Warning: Unbrewed .la files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected .la files:
  /usr/local/lib/libdiskmgt.la
  /usr/local/lib/libnvpair.la
  /usr/local/lib/libuutil.la
  /usr/local/lib/libzfs.la
  /usr/local/lib/libzfs_core.la
  /usr/local/lib/libzpool.la

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libdiskmgt.a
  /usr/local/lib/libnvpair.a
  /usr/local/lib/libuutil.a
  /usr/local/lib/libzfs.a
  /usr/local/lib/libzfs_core.a
  /usr/local/lib/libzpool.a
</pre>

Note:

- The kegs listed by `brew doctor` are all DIY kegs.
- The unbrewed libraries all belong to the OpenZFS cask.

**Comment by jonchang** [comment]
Try unsetting `HOMEBREW_BAT`. What is the output of `brew ruby -e 'puts MacOS.sdk_path_if_needed'`?

**Comment by claui** [comment]
> Try unsetting `HOMEBREW_BAT`.

Just for diagnosis purposes, I assume?
Ok, re-running.

> What is the output of `brew ruby -e 'puts MacOS.sdk_path_if_needed'`?

```
$ brew ruby -e 'puts MacOS.sdk_path_if_needed'
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk
```

**Comment by claui** [comment]
> Try unsetting `HOMEBREW_BAT`.

This makes the following tests pass:

- `Cask::Cmd::Cat given a basic Cask displays the Cask file content about the specified Cask`
- `Cask::Cmd::Cat given a basic Cask can display multiple Casks`
- `brew cat prints the content of a given Formula`
- `brew pin pins a Formula's version`

**Comment by jonchang** [comment]
Homebrew is picking up the 10.15 SDK (presumably from Xcode) instead of the SDK for 10.14 from the CLT. It's weird because that command does point to the right SDK path, but for some reason the tests aren't using that path.

**Comment by claui** [comment]
[5 failures](https://gist.github.com/claui/86fc80e9e451a46c44f4fa244edd9ee8) left.

**Comment by carlocab** [comment]
I get the `tap-new` error too.

I _suspect_ the `pkg-config` errors are because the test spec is outdated. The latest Mojave CLT is actually a Catalina SDK in disguise, that's why the versions the test is finding are the versions for Catalina. Well, except for `expat`; no idea what's going on there. Just a guess though!

**Comment by jonchang** [comment]
The tap-new test failure is due to git 2.28 and https://sfconservancy.org/news/2020/jun/23/gitbranchname/

**Comment by carlocab** [comment]
Ah, I just saw @jonchang's last comment:

>It's weird because that command does point to the right SDK path, but for some reason the tests aren't using that path.

This is because on Mojave:
```
$ ls -l /Library/Developer/CommandLineTools/SDKs/
total 0
drwxr-xr-x 7 root wheel 224 Jul  9  2019 MacOSX.sdk/
lrwxr-xr-x 1 root wheel  10 Dec 20 01:55 MacOSX10.14.sdk -> MacOSX.sdk/
$ jq '.Version' </Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/SDKSettings.json
"10.15"
```

**Comment by Bo98** [comment]
> This is because on Mojave:

What version of the CLT are you using? Installing 11.x over 10.x does that. You must run `sudo rm -rf /Library/Developer/CommandLineTools` before installing 11.x. It's an upgrade bug that was never fixed (and is perhaps why 11.x is not available via `softwareupdate`).

The correct SDK setup on 11.3.1 is:

```
$ ls -l /Library/Developer/CommandLineTools/SDKs
total 0
lrwxr-xr-x 1 root wheel  15 Feb  2 16:38 MacOSX.sdk -> MacOSX10.15.sdk
drwxr-xr-x 7 root wheel 224 Feb  2 16:38 MacOSX10.14.sdk
drwxr-xr-x 8 root wheel 256 Feb  2 16:39 MacOSX10.15.sdk
```

**Comment by carlocab** [comment]
That was from `CLT: 11.3.1.0.1.1576735732`. I think our Mojave CI nodes actually have the same issue.

**Comment by Bo98** [comment]
I will check that. It will need fixing if it is like that. I can't recall if we deliberately left them on 10.3 or updated them - there was a bit of debate at the time.

**Comment by carlocab** [comment]
From `mojave-2`:

```
==> brew config
HOMEBREW_VERSION: 2.7.7-162-g42b755e
ORIGIN: https://github.com/Homebrew/brew
HEAD: 42b755e13f36c34e6a751182f1ec8d7170351549
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 7c26cc9664033bbb10866c93b1891c6adc7992e2
Core tap last commit: 21 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CACHE: /Users/brew/Library/Caches/Homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_COLOR: set
HOMEBREW_DEVELOPER: set
HOMEBREW_FAIL_LOG_LINES: 150
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_GIT_EMAIL: homebrew-test-bot@lists.sfconservancy.org
HOMEBREW_GIT_NAME: BrewTestBot
HOMEBREW_LOGS: /Users/brew/actions-runner/_work/homebrew-core/homebrew-core/logs
HOMEBREW_MAKE_JOBS: 6
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK: set
HOMEBREW_NO_EMOJI: set
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: hexa-core 64-bit penryn
Clang: 11.0 build 1100
Git: 2.21.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.14.6-x86_64
CLT: 11.3.1.0.1.1576735732
Xcode: 11.3.1
```

**Comment by Bo98** [comment]
Ah yes, it's broken:

```
$ ls -l /Library/Developer/CommandLineTools/SDKs/
total 0
lrwxr-xr-x  1 root  wheel   15  5 Dec 12:25 MacOSX.sdk -> MacOSX10.15.sdk
drwxr-xr-x  7 root  wheel  224  5 Dec 12:26 MacOSX.sdk 1
lrwxr-xr-x  1 root  wheel   10 11 Dec  2019 MacOSX10.14.sdk -> MacOSX.sdk
drwxr-xr-x  8 root  wheel  256  7 Nov  2019 MacOSX10.15.sdk
```

I'll get it sorted tomorrow.

**Comment by carlocab** [comment]
Great, would be nice to get rid of workarounds like [this](https://github.com/Homebrew/homebrew-core/blob/4a781c5641f0057b0bf0204149890c656d78cc23/Formula/gcc.rb#L109-L111), [this](https://github.com/Homebrew/homebrew-core/blob/4a781c5641f0057b0bf0204149890c656d78cc23/Formula/gdal.rb#L145-L148), or [this](https://github.com/Homebrew/homebrew-core/blob/4a781c5641f0057b0bf0204149890c656d78cc23/Formula/emacs.rb#L59-L62). (Just leaving notes here so I'll remember where to find them later.)

**Comment by Bo98** [comment]
Oh yikes. I imagine it broke a bunch of Ruby gems as well.

**Comment by Bo98** [comment]
Fixed by #10552 and #10588.

**Comment by claui** [comment]
Thanks so much @carlocab, @Bo98, @jonchang and @reitermarkus! 🍻
